src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs

```diff
@@ -634,7 +634,9 @@ protected override bool MoveNextCore()
                 while (_liveCount < _poolRows && !_doneConsuming)
                 {
                     // We are under capacity. Try to get some more.
-                    while (_toConsumeChannel.Reader.WaitToReadAsync().GetAwaiter().GetResult())
+                    var waitToReadAwaiter = _toConsumeChannel.Reader.WaitToReadAsync().GetAwaiter();
+
+                    while (waitToReadAwaiter.IsCompleted && waitToReadAwaiter.GetResult())
```

Comment:
This is undo-ing what we were trying to avoid with https://github.com/dotnet/machinelearning/pull/5123#discussion_r434105449, which is just spinning a thread seeing if there is something to read.

I was mistaken that you could block synchronously on a Channel. Sorry about that.

What do you think about going back to what you had previously with just calling `TryRead`, and if that returns `false`, the only thing I can think of is to `Thread.Sleep` for some amount of time, even `1 ms`, just to stop the thread from consuming a whole core spinning waiting for the producer. It would probably be a good idea to try it multiple ways and see what perf results you get.

Reply:
Thanks @eerhardt! Made an update for this. Just let me know if it's still not correct.